kubernetes/deployments_kubeclient.go

```diff
@@ -0,0 +1,1339 @@
+	routeURL := fmt.Sprintf("/oapi/v1/namespaces/%s/routes", namespace)
+	result, err := kc.getResource(routeURL, false)
+	if err != nil {
+		return err
+	}
+
+	// Parse list of routes
+	kind, ok := result["kind"].(string)
+	if !ok || kind != "RouteList" {
+		return errors.New("No route list returned from endpoint")
+	}
+	items, ok := result["items"].([]interface{})
+	if !ok {
+		return errors.New("No list of routes in response")
+	}
+
+	for _, item := range items {
+		routeItem, ok := item.(map[interface{}]interface{})
+		if !ok {
+			return errors.New("Route object invalid")
+		}
+
+		// Parse route from result
+		spec, ok := routeItem["spec"].(map[interface{}]interface{})
+		if !ok {
+			return errors.New("Spec missing from route")
+		}
+		// Determine which service this route points to
+		to, ok := spec["to"].(map[interface{}]interface{})
+		if !ok {
+			return errors.New("Route has no destination")
+		}
+		toName, ok := to["name"].(string)
+		if !ok || len(toName) == 0 {
+			return errors.New("Service name missing or invalid for route")
+		}
+
+		var matchingServices []string
+		// Check if this route is for a service we're interested in
+		_, pres := routesByService[toName]
+		if pres {
+			matchingServices = append(matchingServices, toName)
+		}
+
+		// Also check alternate backends for services
+		altBackends, ok := spec["alternateBackends"].([]interface{})
+		if ok {
+			for idx := range altBackends {
+				backend, ok := altBackends[idx].(map[interface{}]interface{})
+				if !ok {
+					return errors.New("Malformed alternative backend")
+				}
+				// Check if this alternate backend is a service we want a route for
+				backendKind, _ := backend["kind"].(string)
+				if backendKind == "Service" {
+					backendName, ok := backend["name"].(string)
+					if ok && len(backendName) > 0 {
+						_, pres := routesByService[backendName]
+						if pres {
+							matchingServices = append(matchingServices, backendName)
+						}
+					}
+				}
+			}
+		}
+		if len(matchingServices) > 0 {
+			// Get ingress points
+			status, ok := routeItem["status"].(map[interface{}]interface{})
+			if !ok {
+				return errors.New("Status missing from route")
+			}
+			ingresses, ok := status["ingress"].([]interface{})
+			if !ok {
+				return errors.New("No ingress array listed in route")
+			}
+
+			// Prefer ingress with oldest lastTransitionTime that is marked as admitted
+			oldestAdmittedIngress, err := findOldestAdmittedIngress(ingresses)
+			if err != nil {
+				return err
+			}
+
+			// Use hostname from oldest admitted ingress if possible
+			var hostname string
+			if oldestAdmittedIngress != nil {
+				hostname, ok = oldestAdmittedIngress["host"].(string)
+				if !ok {
+					return errors.New("Hostname missing from ingress")
+				}
+			} else {
+				// Fall back to optional host in spec
+				hostname, _ = spec["host"].(string)
+			}
+
+			// Check for optional path
+			path, _ := spec["path"].(string)
```

Comment:
are you sure you can skip the type validation ?

Reply:
@xcoulon that sort of thing cause a very ugly bug in our WIT code earlier. Please handle this correctly @stooke !

Reply:
Sure. The problem I was having is that since this is optional, the type-assertion would fail if the key was not present. I'll change this to use two steps, first to check if the key is present and second to perform the type assertion.

Reply:
yes, that's more verbose but that's the way to do it, actually.